apps/frontend/lib/optimization.ts

```
@@ -0,0 +1,501 @@
+
+    search(query: string): number[] {
+        const queryWords = query.toLowerCase().split(/\s+/);
+        const resultSets = queryWords
+            .map(word => this.index.get(word) || new Set())
+            .filter(set => set.size > 0);
+
+        if (resultSets.length === 0) return [];
+
+        // Find intersection of all sets
+        const intersection = resultSets.reduce((acc, set) => {
+            return new Set([...acc].filter(x => set.has(x)));
```

Biome (1.9.4)

[error] Avoid the use of spread (`...`) syntax on accumulators.

Spread syntax should be avoided on accumulators (like those in `.reduce`) because it causes a time complexity of `O(n^2)`.
Consider methods such as .splice or .push instead.

(lint/performance/noAccumulatingSpread)